chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/RealWorldTestingUploadJob.java

```diff
@@ -61,37 +60,64 @@
     public void execute(JobExecutionContext context) throws JobExecutionException {
         SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
         LOGGER.info("********* Starting the Real World Testing Upload job. *********");
-        ExecutorService executorService = null;
 
         try {
             UserDTO user = (UserDTO) context.getMergedJobDataMap().get(USER_KEY);
             setSecurityContext(user);
 
-            Integer threadPoolSize = getThreadCountForJob();
-            executorService = Executors.newFixedThreadPool(threadPoolSize);
+            List<RealWorldTestingUpload> rwts = (List<RealWorldTestingUpload>) context.getMergedJobDataMap()
+                    .get(RWT_UPLOAD_ITEMS);
 
-            List<RealWorldTestingUpload> rwts = (List<RealWorldTestingUpload>) context.getMergedJobDataMap().get(RWT_UPLOAD_ITEMS);
+            // Run some basic validation on upload records that do not have any errors
+            // yet...
+            rwts.stream().filter(rwt -> rwt.getValidationErrors().size() == 0)
+                    .forEach(rwt -> rwt.getValidationErrors().addAll(validateRwtUpload(rwt)));
 
-            List<CompletableFuture<RealWorldTestingUpload>> futures = new ArrayList<CompletableFuture<RealWorldTestingUpload>>();
-            for (RealWorldTestingUpload rwt : rwts) {
-                futures.add(CompletableFuture.supplyAsync(() -> processRwtUploadItem(rwt), executorService));
-            }
+            // Determine if there multiple Plans or Results for the same listing. These will
+            // not get
+            // get processed, and will have an error added to the upload record.
+            rwts = markMultipleChangesForSamePlanTypeAndListing(rwts);
+
+            // Process the plans
+            List<RealWorldTestingUpload> rwtPlans = rwts.stream().filter(
+                    rwt -> rwt.getValidationErrors().size() == 0 && rwt.getType().equals(RealWorldTestingType.PLANS))
+                    .collect(Collectors.toList());
+            saveRealWorldTestingUploads(rwtPlans);
 
-            rwts = futures.stream()
-                    .map(f -> getRwtUploadFromFuture(f))
+            // Process the results
+            List<RealWorldTestingUpload> rwtResults = rwts.stream().filter(
+                    rwt -> rwt.getValidationErrors().size() == 0 && rwt.getType().equals(RealWorldTestingType.RESULTS))
                     .collect(Collectors.toList());
+            saveRealWorldTestingUploads(rwtResults);
```

Comment:
Just so I'm clear, we're still doing asynchronous processing for the plans, and asynchronous processing for the results, but the plans and results, as groups, are done in series, right? So we do all the plans (asynchronously) then when they're done do all the results asynchronously?

Reply:
Correct.